src/reanimated2/animation/spring.ts

```diff
@@ -93,39 +156,123 @@ export function withSpring(
+           * to find the asympotote and esitmate the damping that gives us the expected duration 
+
+                ⎛ ⎛ c⎞           ⎞           
+                ⎜-⎜──⎟ ⋅ duration⎟           
+                ⎝ ⎝2m⎠           ⎠           
+            A ⋅ e                   = treshold
+
+     
+          Amplitude calculated using "Conservation of energy"
+                           _________________
+                          ╱      2         2
+                         ╱ m ⋅ v0  + k ⋅ x0 
+          amplitude =   ╱  ─────────────────
+                      ╲╱           k        
+
+          And replace mass with damping ratio which is provided: m = (c^2)/(4 * k * zeta^2)   
+          */
+      const {
+        duration,
+        velocity: v0,
+        stiffness: k,
+        dampingRatio: zeta,
+        restSpeedThreshold: threshold,
+      } = config;
+
+      const durationForMass = (mass: number) => {
+        const amplitude = Math.sqrt((mass * v0 * v0 + k * x0 * x0) / k);
+        const c = zeta * 2 * Math.sqrt(k * mass);
+        return (
+          1000 * ((-2 * mass) / c) * Math.log((threshold * 0.01) / amplitude) -
+          duration
+        );
+      };
+
+      // Bisection turns out to be much faster than Newton's method in our case
+      return bisectRoot({ min: 0, max: 50, func: durationForMass });
+    }
+
+    function spring(animation: InnerSpringAnimation, now: Timestamp): boolean {
+      const { toValue, startTimestamp, current } = animation;
+
+      const timeFromStart = now - startTimestamp;
+      if (userConfig?.duration && timeFromStart >= userConfig.duration) {
+        animation.current = toValue;
+
+        // clear lastTimestamp to avoid using stale value by the next spring animation that starts after this one
+        animation.lastTimestamp = 0;
+        return true;
       }
 
-      if (isOvershooting() || (isVelocity && isDisplacement)) {
+      const { lastTimestamp, velocity } = animation;
+
+      const deltaTime = Math.min(now - lastTimestamp, 64);
+      animation.lastTimestamp = now;
+
+      const t = deltaTime / 1000;
+      const v0 = -velocity;
+      const x0 = toValue - current;
+
+      const { zeta, omega0, omega1 } = animation;
+
+      const { position: newPosition, velocity: newVelocity } =
+        zeta < 1
+          ? underDampedSpringCalculations(animation, {
+              zeta,
+              v0,
+              x0,
+              omega0,
+              omega1,
+              t,
+            })
+          : criticallyDampedSpringCalculations(animation, {
+              v0,
+              x0,
+              omega0,
+              t,
+            });
+
+      const { isOvershooting, isVelocity, isDisplacement } =
+        isAnimationTerminatingCalculation(animation, config);
+
+      animation.current = newPosition;
+      animation.velocity = newVelocity;
+
+      if (isOvershooting || (isVelocity && isDisplacement)) {
```

Comment:
I think it'd been discussed somewhere before but perhaps got lost: we want for animations with duration set to have additional stop condition such that it ends at the exact moment in time. This is needed because duration could be used to synchronize things happening on screen and we need to make sure the animation finishes exactly when requested

Reply:
I've been thinking you mean cases when the obtained animation is too long and we want to terminate it earlier, but it looks like you want to "animate a static string" as well. IMHO both situations are very different so I've added four options to specify `exactDuration` parameter, instead of a single boolean flag. 🤔